File: ios/Reguerta/Reguerta/Data/ShiftSwapRequests/FirestoreShiftSwapRequestRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

actor FirestoreShiftSwapRequestRepository: ShiftSwapRequestRepository {
    private let storedDB: Firestore
    private let functionsClient: AuthenticatedFirebaseFunctionsClient

    init(firebaseAppName: String, functionsClient: AuthenticatedFirebaseFunctionsClient) {
        guard let app = FirebaseApp.app(name: firebaseAppName) else {
            preconditionFailure("Firebase app is required for shift swap requests")
        }
        self.storedDB = Firestore.firestore(app: app)
        self.functionsClient = functionsClient
    }

    func allShiftSwapRequests(environment: SessionEnvironment) async throws -> [ShiftSwapRequest] {
        let requestsCollection = storedDB.reguertaCollection(.shiftSwapRequests, environment: environment)
        do {
            try Task.checkCancellation()
            let snapshot = try await requestsCollection.getDocuments(source: .server)
            return try snapshot.documents
                .map { document in
                    try Self.shiftSwapRequest(documentID: document.documentID, data: document.data())
                }
                .sorted { $0.requestedAtMillis > $1.requestedAtMillis }
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "shiftSwapRequests")
        }
    }

    func transition(
        _ command: ShiftSwapCommand,
        environment: SessionEnvironment
    ) async throws -> ShiftSwapTransitionResult {
        let request = ShiftSwapTransitionRequest(
            environment: environment,
            command: command
        )
        let response: ShiftSwapTransitionResponse
        do {
            try Task.checkCancellation()
            response = try await functionsClient.post(
                function: .transitionShiftSwap,
                body: request,
                response: ShiftSwapTransitionResponse.self
            )
        } catch {
            throw Self.mapFunctionError(error)
        }
        let responseRequestId = response.requestId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard response.ok,
              response.environment == request.environment,
              response.action == request.action,
              !responseRequestId.isEmpty,
              request.requestId == nil || request.requestId == responseRequestId else {
            throw ShiftSwapCommandError.invalidData
        }
        return ShiftSwapTransitionResult(
            requestId: responseRequestId,
            candidateCount: response.candidateCount
        )
    }

    static func shiftSwapRequest(documentID: String, data: [String: Any]) throws -> ShiftSwapRequest {
        guard !documentID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let requestedShiftId = nonEmptyTrimmedString(data["requestedShiftId"]),
              let requesterUserId = nonEmptyTrimmedString(data["requesterUserId"]),
              let reason = data["reason"] as? String,
              let statusRaw = nonEmptyTrimmedString(data["status"])?.lowercased(),
              let status = ShiftSwapRequestStatus(rawValue: statusRaw),
              let rawCandidates = data["candidates"] as? [[String: Any]],
              let rawResponses = data["responses"] as? [[String: Any]],
              let requestedAt = data["requestedAt"] as? Timestamp else {
            throw invalidDocumentError
        }

        let candidates = try rawCandidates.map(Self.shiftSwapCandidate)
        let responses = try rawResponses.map(Self.shiftSwapResponse)

        return ShiftSwapRequest(
            id: documentID,
            requestedShiftId: requestedShiftId,
            requesterUserId: requesterUserId,
            reason: reason.trimmingCharacters(in: .whitespacesAndNewlines),
            status: status,
            candidates: candidates,
            responses: responses,
            selectedCandidateUserId: try optionalNonEmptyString(data["selectedCandidateUserId"]),
            selectedCandidateShiftId: try optionalNonEmptyString(data["selectedCandidateShiftId"]),
            requestedAtMillis: Int64(requestedAt.dateValue().timeIntervalSince1970 * 1_000),
            confirmedAtMillis: try optionalTimestampMillis(data["confirmedAt"]),
            appliedAtMillis: try optionalTimestampMillis(data["appliedAt"])
        )
    }

    private static func shiftSwapCandidate(_ item: [String: Any]) throws -> ShiftSwapCandidate {
        guard let userId = nonEmptyTrimmedString(item["userId"]),
              let shiftId = nonEmptyTrimmedString(item["shiftId"]) else {
            throw invalidDocumentError
        }
        return ShiftSwapCandidate(userId: userId, shiftId: shiftId)
    }

    private static func shiftSwapResponse(_ item: [String: Any]) throws -> ShiftSwapResponse {
        guard let userId = nonEmptyTrimmedString(item["userId"]),
              let shiftId = nonEmptyTrimmedString(item["shiftId"]),
              let statusRaw = nonEmptyTrimmedString(item["status"])?.lowercased(),
              let status = ShiftSwapResponseStatus(rawValue: statusRaw),
              let respondedAt = item["respondedAt"] as? Timestamp else {
            throw invalidDocumentError
        }
        return ShiftSwapResponse(
            userId: userId,
            shiftId: shiftId,
            status: status,
            respondedAtMillis: Int64(respondedAt.dateValue().timeIntervalSince1970 * 1_000)
        )
    }

    private static func nonEmptyTrimmedString(_ value: Any?) -> String? {
        guard let text = value as? String else { return nil }
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func optionalNonEmptyString(_ value: Any?) throws -> String? {
        guard let value else { return nil }
        if value is NSNull { return nil }
        guard let parsed = nonEmptyTrimmedString(value) else { throw invalidDocumentError }
        return parsed
    }

    private static func optionalTimestampMillis(_ value: Any?) throws -> Int64? {
        guard let value else { return nil }
        if value is NSNull { return nil }
        guard let timestamp = value as? Timestamp else { throw invalidDocumentError }
        return Int64(timestamp.dateValue().timeIntervalSince1970 * 1_000)
    }

    private static func mapFunctionError(_ error: any Error) -> any Error {
        if error is CancellationError {
            return error
        }
        if let commandError = error as? ShiftSwapCommandError {
            return commandError
        }
        if let repositoryError = error as? RepositoryError {
            return mapRepositoryError(repositoryError)
        }
        guard let functionError = error as? FirebaseFunctionClientError else {
            return ShiftSwapCommandError.unknown
        }
        return mapFunctionClientError(functionError)
    }

    private static func mapRepositoryError(_ error: RepositoryError) -> ShiftSwapCommandError {
        switch error {
        case .notFound:
            return .conflict(code: "not_found")
        case .unavailable:
            return .unavailable
        case .permissionDenied:
            return .permissionDenied
        case .invalidData:
            return .invalidData
        case .unknown:
            return .unknown
        }
    }

    private static func mapFunctionClientError(_ error: FirebaseFunctionClientError) -> any Error {
        switch error {
        case .cancelled:
            return CancellationError()
        case .missingIDToken, .unauthorized, .forbidden:
            return ShiftSwapCommandError.permissionDenied
        case .conflict(let code, _):
            return code == "no_shift_swap_candidates"
                ? ShiftSwapCommandError.noCandidates
                : ShiftSwapCommandError.conflict(code: code)
        case .timeout, .transport:
            return ShiftSwapCommandError.unavailable
        case .http(let statusCode, let code, _):
            return mapFunctionHTTPError(statusCode: statusCode, code: code)
        case .invalidHTTPResponse, .invalidResponse:
            return ShiftSwapCommandError.invalidData
        case .invalidEndpoint:
            return ShiftSwapCommandError.unknown
        }
    }

    private static func mapFunctionHTTPError(statusCode: Int, code: String) -> ShiftSwapCommandError {
        if statusCode == 408 || statusCode == 429 || statusCode >= 500 {
            return .unavailable
        }
        if statusCode == 400 {
            return .invalidData
        }
        if statusCode == 404 {
            return .conflict(code: code)
        }
        return .unknown
    }

    private static var invalidDocumentError: RepositoryError {
        .invalidData(resource: "shiftSwapRequests.document")
    }
}

nonisolated private struct ShiftSwapTransitionResponse: Decodable {
    let ok: Bool
    let environment: SessionEnvironment
    let action: String
    let requestId: String
    let candidateCount: Int?
}

nonisolated private struct ShiftSwapTransitionRequest: Encodable {
    let environment: SessionEnvironment
    let action: String
    let requestedShiftId: String?
    let reason: String?
    let requestId: String?
    let candidateShiftId: String?
    let response: String?
}

extension ShiftSwapTransitionRequest {
    init(environment: SessionEnvironment, command: ShiftSwapCommand) {
        self.environment = environment
        switch command {
        case .create(let requestedShiftId, let reason):
            action = "create"
            self.requestedShiftId = requestedShiftId
            self.reason = reason
            requestId = nil
            candidateShiftId = nil
            response = nil
        case .respond(let requestId, let candidateShiftId, let responseStatus):
            action = "respond"
            requestedShiftId = nil
            reason = nil
            self.requestId = requestId
            self.candidateShiftId = candidateShiftId
            response = responseStatus.rawValue
        case .cancel(let requestId):
            action = "cancel"
            requestedShiftId = nil
            reason = nil
            self.requestId = requestId
            candidateShiftId = nil
            response = nil
        case .apply(let requestId, let candidateShiftId):
            action = "apply"
            requestedShiftId = nil
            reason = nil
            self.requestId = requestId
            self.candidateShiftId = candidateShiftId
            response = nil
        }
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/shiftswap/FirestoreShiftSwapRequestRepository.kt
package com.reguerta.user.data.shiftswap

import com.google.android.gms.tasks.Tasks
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Source
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.shifts.ShiftSwapRequest
import com.reguerta.user.domain.shifts.ShiftSwapCandidate
import com.reguerta.user.domain.shifts.ShiftSwapRequestRepository
import com.reguerta.user.domain.shifts.ShiftSwapRequestStatus
import com.reguerta.user.domain.shifts.ShiftSwapResponse
import com.reguerta.user.domain.shifts.ShiftSwapResponseStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

internal class FirestoreShiftSwapRequestRepository(
    private val firestore: FirebaseFirestore,
    private val transitionClient: FirebaseShiftSwapTransitionClient,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : ShiftSwapRequestRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val requestsCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.SHIFT_SWAP_REQUESTS)

    override suspend fun getAllShiftSwapRequests(): List<ShiftSwapRequest> = withContext(Dispatchers.IO) {
        try {
            val snapshot = Tasks.await(
                firestore.collection(requestsCollectionPath).get(Source.SERVER),
            )
            decodeShiftSwapRequestDocuments(
                snapshot.documents.map { document ->
                    document.id to (document.data ?: invalidShiftSwapRequestDocument())
                },
            )
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "shiftSwapRequests")
        }
    }

    override suspend fun createShiftSwapRequest(requestedShiftId: String, reason: String): String =
        try {
            transitionClient.create(requestedShiftId = requestedShiftId, reason = reason)
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "shiftSwapRequests.transition")
        }

    override suspend fun respondToShiftSwapRequest(
        requestId: String,
        candidateShiftId: String,
        response: ShiftSwapResponseStatus,
    ) {
        try {
            transitionClient.respond(
                requestId = requestId,
                candidateShiftId = candidateShiftId,
                response = response,
            )
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "shiftSwapRequests.transition")
        }
    }

    override suspend fun cancelShiftSwapRequest(requestId: String) {
        try {
            transitionClient.cancel(requestId)
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "shiftSwapRequests.transition")
        }
    }

    override suspend fun applyShiftSwapRequest(requestId: String, candidateShiftId: String) {
        try {
            transitionClient.apply(requestId = requestId, candidateShiftId = candidateShiftId)
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "shiftSwapRequests.transition")
        }
    }
}

internal fun decodeShiftSwapRequestDocuments(
    documents: List<Pair<String, Map<String, Any?>>>,
): List<ShiftSwapRequest> = documents
    .map { (documentId, data) -> decodeShiftSwapRequestDocument(documentId, data) }
    .sortedByDescending { it.requestedAtMillis }

private fun decodeShiftSwapRequestDocument(
    documentId: String,
    data: Map<String, Any?>,
): ShiftSwapRequest {
    val requestedShiftId = data.requiredString("requestedShiftId")
    val requesterUserId = data.requiredString("requesterUserId")
    val reason = data.requiredPossiblyEmptyString("reason")
    val status = data.requiredString("status").lowercase().toShiftSwapRequestStatus()
        ?: invalidShiftSwapRequestDocument()
    val requestedAtMillis = data.requiredTimestampMillis("requestedAt")
    val candidates = data.requiredObjectList("candidates").map { candidate ->
        ShiftSwapCandidate(
            userId = candidate.requiredString("userId"),
            shiftId = candidate.requiredString("shiftId"),
        )
    }
    val responses = data.requiredObjectList("responses").map { response ->
        val responseStatus = response.requiredString("status").lowercase().toShiftSwapResponseStatus()
            ?: invalidShiftSwapRequestDocument()
        ShiftSwapResponse(
            userId = response.requiredString("userId"),
            shiftId = response.requiredString("shiftId"),
            status = responseStatus,
            respondedAtMillis = response.requiredTimestampMillis("respondedAt"),
        )
    }
    return ShiftSwapRequest(
        id = documentId,
        requestedShiftId = requestedShiftId,
        requesterUserId = requesterUserId,
        reason = reason,
        status = status,
        candidates = candidates,
        responses = responses,
        selectedCandidateUserId = data.optionalString("selectedCandidateUserId"),
        selectedCandidateShiftId = data.optionalString("selectedCandidateShiftId"),
        requestedAtMillis = requestedAtMillis,
        confirmedAtMillis = data.optionalTimestampMillis("confirmedAt"),
        appliedAtMillis = data.optionalTimestampMillis("appliedAt"),
    )
}

private fun Map<String, Any?>.requiredString(field: String): String =
    optionalString(field) ?: invalidShiftSwapRequestDocument()

private fun Map<String, Any?>.requiredPossiblyEmptyString(field: String): String {
    if (!containsKey(field)) invalidShiftSwapRequestDocument()
    return (this[field] as? String)?.trim() ?: invalidShiftSwapRequestDocument()
}

private fun Map<String, Any?>.optionalString(field: String): String? {
    if (!containsKey(field) || this[field] == null) return null
    val value = this[field] as? String ?: invalidShiftSwapRequestDocument()
    return value.trim().ifBlank { null }
}

private fun Map<String, Any?>.requiredObjectList(field: String): List<Map<String, Any?>> {
    val values = this[field] as? List<*> ?: invalidShiftSwapRequestDocument()
    return values.map { value -> value.toStringKeyedMap() }
}

private fun Any?.toStringKeyedMap(): Map<String, Any?> {
    val raw = this as? Map<*, *> ?: invalidShiftSwapRequestDocument()
    return raw.entries.associate { (key, value) ->
        (key as? String ?: invalidShiftSwapRequestDocument()) to value
    }
}

private fun Map<String, Any?>.requiredTimestampMillis(field: String): Long =
    optionalTimestampMillis(field) ?: invalidShiftSwapRequestDocument()

private fun Map<String, Any?>.optionalTimestampMillis(field: String): Long? {
    if (!containsKey(field) || this[field] == null) return null
    return (this[field] as? Timestamp)?.toDate()?.time ?: invalidShiftSwapRequestDocument()
}

private fun invalidShiftSwapRequestDocument(): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "shiftSwapRequests.document",
)

private fun String.toShiftSwapRequestStatus(): ShiftSwapRequestStatus? = when (this) {
    "open" -> ShiftSwapRequestStatus.OPEN
    "cancelled" -> ShiftSwapRequestStatus.CANCELLED
    "applied" -> ShiftSwapRequestStatus.APPLIED
    else -> null
}

private fun String.toShiftSwapResponseStatus(): ShiftSwapResponseStatus? = when (this) {
    "available" -> ShiftSwapResponseStatus.AVAILABLE
    "unavailable" -> ShiftSwapResponseStatus.UNAVAILABLE
    else -> null
}
